File: packages/shared/lib/utils/plotCacheService.ts
import logger from "./logger";

export interface PlotParams {
  chunk_start: number;
  chunk_size: number;
  preprocessing_options?: Record<string, any>;
  selected_channels?: string[];
  time_window?: [number, number];
  zoom_level?: number;
}

export interface CachePlotRequest {
  file_path: string;
  plot_params: PlotParams;
  plot_data: Record<string, any>;
  ttl?: number;
}

export interface CachePlotResponse {
  success: boolean;
  message: string;
  cache_key?: string;
}

export interface GetCachedPlotRequest {
  file_path: string;
  plot_params: PlotParams;
}

export interface GetCachedPlotResponse {
  success: boolean;
  message: string;
  plot_data?: Record<string, any>;
  cached_at?: string;
}

export interface CachedPlotMetadata {
  cache_key: string;
  file_path: string;
  plot_params: PlotParams;
  cached_at: string;
  ttl: number;
}

export interface UserCachedPlotsResponse {
  success: boolean;
  message: string;
  plots: CachedPlotMetadata[];
  total_count: number;
}

export interface DeleteCachedPlotRequest {
  file_path: string;
  plot_params: PlotParams;
}

export interface DeleteCachedPlotResponse {
  success: boolean;
  message: string;
  deleted: boolean;
}

export interface DeleteFilePlotsRequest {
  file_path: string;
}

export interface DeleteFilePlotsResponse {
  success: boolean;
  message: string;
  deleted_count: number;
}

export interface DeleteUserPlotsResponse {
  success: boolean;
  message: string;
  deleted_count: number;
}

export interface CleanupResponse {
  success: boolean;
  message: string;
  cleaned_count: number;
}

class PlotCacheService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = "/api/plot-cache";
  }

  private async makeRequest<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    try {
      const url = `${this.baseUrl}${endpoint}`;
      const response = await fetch(url, {
        ...options,
        headers: {
          "Content-Type": "application/json",
          ...options.headers,
        },
        credentials: "include",
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      return data as T;
    } catch (error) {
      logger.error(`PlotCacheService error for ${endpoint}:`, error);
      throw error;
    }
  }

  /**
   * Cache a plot for the current user
   */
  async cachePlot(request: CachePlotRequest): Promise<CachePlotResponse> {
    logger.info(
      `[PlotCacheService] Caching plot for file: ${request.file_path}`
    );
    logger.debug(`[PlotCacheService] Cache request details:`, {
      file_path: request.file_path,
      plot_params: request.plot_params,
      ttl: request.ttl,
      plot_data_keys: Object.keys(request.plot_data || {}),
    });

    try {
      const response = await this.makeRequest<CachePlotResponse>("/cache", {
        method: "POST",
        body: JSON.stringify(request),
      });

      logger.info(
        `[PlotCacheService] Cache response: ${
          response.success ? "SUCCESS" : "FAILED"
        } - ${response.message}`
      );
      return response;
    } catch (error) {
      logger.error(`[PlotCacheService] Cache request failed:`, error);
      throw error;
    }
  }

  /**
   * Retrieve a cached plot for the current user
   */
  async getCachedPlot(
    request: GetCachedPlotRequest
  ): Promise<GetCachedPlotResponse> {
    logger.info(
      `[PlotCacheService] Getting cached plot for file: ${request.file_path}`
    );
    logger.debug(`[PlotCacheService] Get request details:`, {
      file_path: request.file_path,
      plot_params: request.plot_params,
    });

    try {
      const response = await this.makeRequest<GetCachedPlotResponse>("/get", {
        method: "POST",
        body: JSON.stringify(request),
      });

      logger.info(
        `[PlotCacheService] Get response: ${
          response.success ? "FOUND" : "NOT_FOUND"
        } - ${response.message}`
      );
      if (response.success && response.plot_data) {
        logger.debug(
          `[PlotCacheService] Retrieved plot data keys:`,
          Object.keys(response.plot_data)
        );
      }
      return response;
    } catch (error) {
      logger.error(`[PlotCacheService] Get request failed:`, error);
      throw error;
    }
  }

  /**
   * Get all cached plots for the current user
   */
  async getUserCachedPlots(): Promise<UserCachedPlotsResponse> {
    logger.info(`[PlotCacheService] Getting all cached plots for user`);

    try {
      const response = await this.makeRequest<UserCachedPlotsResponse>(
        "/user",
        {
          method: "GET",
        }
      );

      logger.info(
        `[PlotCacheService] User plots response: ${
          response.success ? "SUCCESS" : "FAILED"
        } - ${response.total_count} plots`
      );
      if (response.success && response.plots) {
        response.plots.forEach((plot) => {
          logger.debug(
            `[PlotCacheService] User plot: ${plot.file_path} - ${plot.cached_at}`
          );
        });
      }
      return response;
    } catch (error) {
      logger.error(`[PlotCacheService] Get user plots request failed:`, error);
      throw error;
    }
  }

  /**
   * Delete a specific cached plot for the current user
   */
  async deleteCachedPlot(
    request: DeleteCachedPlotRequest
  ): Promise<DeleteCachedPlotResponse> {
    return this.makeRequest<DeleteCachedPlotResponse>("/delete", {
      method: "DELETE",
      body: JSON.stringify(request),
    });
  }

  /**
   * Delete all cached plots for a specific file for the current user
   */
  async deleteFilePlots(
    request: DeleteFilePlotsRequest
  ): Promise<DeleteFilePlotsResponse> {
    return this.makeRequest<DeleteFilePlotsResponse>("/delete-file", {
      method: "DELETE",
      body: JSON.stringify(request),
    });
  }

  /**
   * Delete all cached plots for the current user
   */
  async deleteUserPlots(): Promise<DeleteUserPlotsResponse> {
    return this.makeRequest<DeleteUserPlotsResponse>("/delete-user", {
      method: "DELETE",
    });
  }

  /**
   * Clean up expired plots for all users (admin operation)
   */
  async cleanupExpiredPlots(): Promise<CleanupResponse> {
    return this.makeRequest<CleanupResponse>("/cleanup", {
      method: "POST",
    });
  }

  /**
   * Check if a cached plot exists for the given parameters
   */
  async hasCachedPlot(
    filePath: string,
    plotParams: PlotParams
  ): Promise<boolean> {
    try {
      const response = await this.getCachedPlot({
        file_path: filePath,
        plot_params: plotParams,
      });
      return response.success && response.plot_data !== undefined;
    } catch (error) {
      logger.error("Error checking for cached plot:", error);
      return false;
    }
  }

  /**
   * Load cached plots for a user and return them in a format compatible with the Redux store
   */
  async loadUserCachedPlots(): Promise<Record<string, any>> {
    logger.info(`[PlotCacheService] Starting loadUserCachedPlots`);

    try {
      const response = await this.getUserCachedPlots();
      if (!response.success) {
        logger.warn(
          `[PlotCacheService] Failed to load user cached plots: ${response.message}`
        );
        return {};
      }

      logger.info(
        `[PlotCacheService] Found ${response.plots.length} cached plots to restore`
      );

      // Convert cached plots to Redux store format
      const plots: Record<string, any> = {};

      for (const plotMetadata of response.plots) {
        logger.debug(
          `[PlotCacheService] Processing plot: ${plotMetadata.file_path}`
        );

        const plotResponse = await this.getCachedPlot({
          file_path: plotMetadata.file_path,
          plot_params: plotMetadata.plot_params,
        });

        if (plotResponse.success && plotResponse.plot_data) {
          // Create a unique key for the plot in Redux store
          const plotKey = `${plotMetadata.file_path}_${plotMetadata.plot_params.chunk_start}_${plotMetadata.plot_params.chunk_size}`;
          logger.debug(`[PlotCacheService] Creating plot key: ${plotKey}`);

          plots[plotKey] = {
            ...plotResponse.plot_data,
            filePath: plotMetadata.file_path,
            chunkStart: plotMetadata.plot_params.chunk_start,
            chunkSizeSeconds: plotMetadata.plot_params.chunk_size,
            preprocessingOptions:
              plotMetadata.plot_params.preprocessing_options,
            selectedChannels: plotMetadata.plot_params.selected_channels,
            timeWindow: plotMetadata.plot_params.time_window,
            zoomLevel: plotMetadata.plot_params.zoom_level,
            cachedAt: plotMetadata.cached_at,
          };

          logger.debug(
            `[PlotCacheService] Restored plot: ${plotKey} with keys:`,
            Object.keys(plots[plotKey])
          );
        } else {
          logger.warn(
            `[PlotCacheService] Failed to retrieve plot data for: ${plotMetadata.file_path}`
          );
        }
      }

      logger.info(
        `[PlotCacheService] Successfully loaded ${
          Object.keys(plots).length
        } cached plots from Redis`
      );
      return plots;
    } catch (error) {
      logger.error(
        `[PlotCacheService] Error loading user cached plots:`,
        error
      );
      return {};
    }
  }
}

// Export singleton instance
export const plotCacheService = new PlotCacheService();

File: packages/api/core/services/plot_cache_service.py
"""Plot caching service using Redis."""

import hashlib
import json
from datetime import datetime
from typing import Any, Dict, List, Optional

from loguru import logger

from ..environment import get_config_service
from ..service_registry import register_service
from .base import BaseService
from .redis_service import RedisService


@register_service
class PlotCacheService(BaseService):
    """Service for caching plot data in Redis."""

    def __init__(self):
        super().__init__(None)  # No database dependency needed
        self.cache_settings = get_config_service().get_cache_settings()
        self.redis_service = RedisService(
            host=self.cache_settings.redis_host,
            port=self.cache_settings.redis_port,
            db=self.cache_settings.redis_db,
            password=self.cache_settings.redis_password,
            use_ssl=self.cache_settings.redis_use_ssl,
        )

    @classmethod
    def from_db(cls, db=None) -> "PlotCacheService":
        """Create a service instance for dependency injection."""
        return cls()

    def _generate_cache_key(
        self, user_id: int, file_path: str, plot_params: Dict[str, Any]
    ) -> str:
        """Generate a unique cache key for plot data."""
        # Create a hash of the plot parameters to ensure uniqueness
        params_hash = hashlib.md5(
            json.dumps(plot_params, sort_keys=True).encode()
        ).hexdigest()

        # Create a safe file path hash
        file_path_hash = hashlib.md5(file_path.encode()).hexdigest()

        cache_key = (
            f"plot_cache:user:{user_id}:file:{file_path_hash}:params:{params_hash}"
        )
        logger.debug(f"[PlotCacheService] Generated cache key: {cache_key}")
        logger.debug(
            f"[PlotCacheService] Cache key components - user_id: {user_id}, file_path: {file_path}, params_hash: {params_hash}"
        )
        return cache_key

    def _generate_user_plots_key(self, user_id: int) -> str:
        """Generate a key for storing user's plot cache metadata."""
        user_plots_key = f"plot_cache:user:{user_id}:plots"
        logger.debug(f"[PlotCacheService] Generated user plots key: {user_plots_key}")
        return user_plots_key

    async def cache_plot(
        self,
        user_id: int,
        file_path: str,
        plot_params: Dict[str, Any],
        plot_data: Dict[str, Any],
        ttl: Optional[int] = None,
    ) -> bool:
        """
        Cache plot data for a user.

        Args:
            user_id: The user ID
            file_path: The file path
            plot_params: Plot parameters (chunk_start, chunk_size, preprocessing_options, etc.)
            plot_data: The actual plot data to cache
            ttl: Time to live in seconds (uses default from settings if None)

        Returns:
            True if successful, False otherwise
        """
        logger.info(
            f"[PlotCacheService] Starting cache_plot for user {user_id}, file {file_path}"
        )
        logger.debug(f"[PlotCacheService] Plot params: {plot_params}")
        logger.debug(
            f"[PlotCacheService] Plot data keys: {list(plot_data.keys()) if plot_data else 'None'}"
        )

        try:
            cache_key = self._generate_cache_key(user_id, file_path, plot_params)
            user_plots_key = self._generate_user_plots_key(user_id)

            # Use default TTL if not specified
            if ttl is None:
                ttl = self.cache_settings.plot_cache_ttl
                logger.debug(f"[PlotCacheService] Using default TTL: {ttl}s")
            else:
                logger.debug(f"[PlotCacheService] Using provided TTL: {ttl}s")

            # Cache the plot data
            logger.debug(f"[PlotCacheService] Caching plot data with key: {cache_key}")
            success = await self.redis_service.set(cache_key, plot_data, ttl)
            if not success:
                logger.error(
                    f"[PlotCacheService] Failed to cache plot data for user {user_id}, file {file_path}"
                )
                return False

            # Store metadata about this cached plot
            plot_metadata = {
                "cache_key": cache_key,
                "file_path": file_path,
                "plot_params": plot_params,
                "cached_at": datetime.utcnow().isoformat(),
                "ttl": ttl,
            }

            logger.debug(f"[PlotCacheService] Storing plot metadata: {plot_metadata}")

            # Get existing user plots metadata
            user_plots = await self.redis_service.get_json(user_plots_key) or {}
            logger.debug(
                f"[PlotCacheService] Existing user plots count: {len(user_plots)}"
            )

            # Add or update the plot metadata
            user_plots[cache_key] = plot_metadata

            # Store updated user plots metadata
            await self.redis_service.set_json(user_plots_key, user_plots, ttl)

            logger.info(
                f"[PlotCacheService] Successfully cached plot for user {user_id}, file {file_path}, key: {cache_key}"
            )
            logger.debug(
                f"[PlotCacheService] Total cached plots for user {user_id}: {len(user_plots)}"
            )
            return True

        except Exception as e:
            logger.error(
                f"[PlotCacheService] Error caching plot for user {user_id}, file {file_path}: {e}"
            )
            return False

    async def get_cached_plot(
        self, user_id: int, file_path: str, plot_params: Dict[str, Any]
    ) -> Optional[Dict[str, Any]]:
        """
        Retrieve cached plot data for a user.

        Args:
            user_id: The user ID
            file_path: The file path
            plot_params: Plot parameters (chunk_start, chunk_size, preprocessing_options, etc.)

        Returns:
            Cached plot data if found, None otherwise
        """
        logger.info(
            f"[PlotCacheService] Starting get_cached_plot for user {user_id}, file {file_path}"
        )
        logger.debug(f"[PlotCacheService] Plot params: {plot_params}")

        try:
            cache_key = self._generate_cache_key(user_id, file_path, plot_params)
            logger.debug(f"[PlotCacheService] Looking for cache key: {cache_key}")

            # Check if the cached plot exists
            exists = await self.redis_service.exists(cache_key)
            if not exists:
                logger.info(
                    f"[PlotCacheService] No cached plot found for user {user_id}, file {file_path}"
                )
                return None

            # Get the cached plot data
            logger.debug("[PlotCacheService] Cached plot exists, retrieving data...")
            plot_data = await self.redis_service.get(cache_key)
            if plot_data is None:
                logger.warning(
                    f"[PlotCacheService] Failed to retrieve cached plot data for key: {cache_key}"
                )
                return None

            logger.info(
                f"[PlotCacheService] Successfully retrieved cached plot for user {user_id}, file {file_path}"
            )
            logger.debug(
                f"[PlotCacheService] Retrieved plot data keys: {list(plot_data.keys()) if plot_data else 'None'}"
            )
            return plot_data

        except Exception as e:
            logger.error(
                f"[PlotCacheService] Error retrieving cached plot for user {user_id}, file {file_path}: {e}"
            )
            return None

    async def get_user_cached_plots(self, user_id: int) -> List[Dict[str, Any]]:
        """
        Get all cached plots for a user.

        Args:
            user_id: The user ID

        Returns:
            List of cached plot metadata
        """
        logger.info(
            f"[PlotCacheService] Starting get_user_cached_plots for user {user_id}"
        )

        try:
            user_plots_key = self._generate_user_plots_key(user_id)
            logger.debug(
                f"[PlotCacheService] Getting user plots from key: {user_plots_key}"
            )

            user_plots = await self.redis_service.get_json(user_plots_key) or {}
            logger.debug(
                f"[PlotCacheService] Found {len(user_plots)} cached plots for user {user_id}"
            )

            # Filter out expired plots
            valid_plots = []
            expired_count = 0

            for cache_key, plot_metadata in user_plots.items():
                logger.debug(f"[PlotCacheService] Checking plot: {cache_key}")
                exists = await self.redis_service.exists(cache_key)
                if exists:
                    valid_plots.append(plot_metadata)
                    logger.debug(f"[PlotCacheService] Plot is valid: {cache_key}")
                else:
                    # Remove expired plot from metadata
                    logger.debug(f"[PlotCacheService] Plot has expired: {cache_key}")
                    del user_plots[cache_key]
                    expired_count += 1

            # Update user plots metadata if any expired plots were removed
            if expired_count > 0:
                logger.info(
                    f"[PlotCacheService] Removed {expired_count} expired plots for user {user_id}"
                )
                await self.redis_service.set_json(user_plots_key, user_plots)

            logger.info(
                f"[PlotCacheService] Retrieved {len(valid_plots)} valid cached plots for user {user_id}"
            )
            for plot in valid_plots:
                logger.debug(
                    f"[PlotCacheService] Valid plot: {plot.get('file_path', 'unknown')} - {plot.get('cached_at', 'unknown')}"
                )

            return valid_plots

        except Exception as e:
            logger.error(
                f"[PlotCacheService] Error retrieving cached plots for user {user_id}: {e}"
            )
            return []

    async def delete_cached_plot(
        self, user_id: int, file_path: str, plot_params: Dict[str, Any]
    ) -> bool:
        """
        Delete a specific cached plot.

        Args:
            user_id: The user ID
            file_path: The file path
            plot_params: Plot parameters

        Returns:
            True if successful, False otherwise
        """
        try:
            cache_key = self._generate_cache_key(user_id, file_path, plot_params)
            user_plots_key = self._generate_user_plots_key(user_id)

            # Delete the cached plot data
            success = await self.redis_service.delete(cache_key)

            # Remove from user plots metadata
            user_plots = await self.redis_service.get_json(user_plots_key) or {}
            if cache_key in user_plots:
                del user_plots[cache_key]
                await self.redis_service.set_json(user_plots_key, user_plots)

            if success:
                logger.info(f"Deleted cached plot for user {user_id}, file {file_path}")
            else:
                logger.warning(
                    f"No cached plot found to delete for user {user_id}, file {file_path}"
                )

            return success

        except Exception as e:
            logger.error(
                f"Error deleting cached plot for user {user_id}, file {file_path}: {e}"
            )
            return False

    async def delete_user_plots(self, user_id: int) -> bool:
        """
        Delete all cached plots for a user.

        Args:
            user_id: The user ID

        Returns:
            True if successful, False otherwise
        """
        try:
            user_plots_key = self._generate_user_plots_key(user_id)
            user_plots = await self.redis_service.get_json(user_plots_key) or {}

            # Delete all cached plot data
            deleted_count = 0
            for cache_key in user_plots.keys():
                if await self.redis_service.delete(cache_key):
                    deleted_count += 1

            # Delete user plots metadata
            await self.redis_service.delete(user_plots_key)

            logger.info(f"Deleted {deleted_count} cached plots for user {user_id}")
            return True

        except Exception as e:
            logger.error(f"Error deleting cached plots for user {user_id}: {e}")
            return False

    async def delete_file_plots(self, user_id: int, file_path: str) -> bool:
        """
        Delete all cached plots for a specific file.

        Args:
            user_id: The user ID
            file_path: The file path

        Returns:
            True if successful, False otherwise
        """
        try:
            user_plots_key = self._generate_user_plots_key(user_id)
            user_plots = await self.redis_service.get_json(user_plots_key) or {}

            # Find and delete all plots for this file
            deleted_count = 0
            keys_to_remove = []

            for cache_key, plot_metadata in user_plots.items():
                if plot_metadata.get("file_path") == file_path:
                    if await self.redis_service.delete(cache_key):
                        deleted_count += 1
                    keys_to_remove.append(cache_key)

            # Remove from user plots metadata
            for key in keys_to_remove:
                del user_plots[key]

            if keys_to_remove:
                await self.redis_service.set_json(user_plots_key, user_plots)

            logger.info(
                f"Deleted {deleted_count} cached plots for user {user_id}, file {file_path}"
            )
            return True

        except Exception as e:
            logger.error(
                f"Error deleting cached plots for user {user_id}, file {file_path}: {e}"
            )
            return False

    async def cleanup_expired_plots(self) -> int:
        """
        Clean up expired plots from all users.

        Returns:
            Number of expired plots cleaned up
        """
        try:
            # Get all user plot metadata keys
            user_keys = await self.redis_service.get_keys("plot_cache:user:*:plots")

            total_cleaned = 0

            for user_key in user_keys:
                user_plots = await self.redis_service.get_json(user_key) or {}

                # Check each plot for expiration
                keys_to_remove = []
                for cache_key, plot_metadata in user_plots.items():
                    exists = await self.redis_service.exists(cache_key)
                    if not exists:
                        keys_to_remove.append(cache_key)
                        total_cleaned += 1

                # Remove expired plots from metadata
                for key in keys_to_remove:
                    del user_plots[key]

                # Update user plots metadata if any expired plots were removed
                if keys_to_remove:
                    await self.redis_service.set_json(user_key, user_plots)

            logger.info(f"Cleaned up {total_cleaned} expired cached plots")
            return total_cleaned

        except Exception as e:
            logger.error(f"Error cleaning up expired plots: {e}")
            return 0

    async def health_check(self) -> bool:
        """Check if the service is healthy."""
        try:
            # Check if Redis connection is working by trying a simple operation
            await self.redis_service.exists("health_check_test")
            return True
        except Exception as e:
            logger.error(f"PlotCacheService health check failed: {e}")
            return False

    async def close(self):
        """Close Redis connection."""
        await self.redis_service.close()
